vendor/noise_handshake/src/handshake.rs

```diff
@@ -0,0 +1,230 @@
+use futures::*;
+use snow::Session::{self, Handshake};
+use tokio::io::{AsyncRead, AsyncWrite};
+
+const NOISE_MAX_SIZE: usize = 65535;
```

Comment:
Do we actually need that much for the handshake? Instances of the enum will have quite a big memory footprint with this size.

Reply:
Hum, no we apparently do not but we would need to make assumptions.

Reply:
The way you've implemented this you will be able to handle any noise handshake successfully! So well done on that. It's possible you will need a buffer of 65535 in case the application puts a lot of data on top of the handshake messages.

The thing is, you're completely dropping all the application data below:

```rust
                let mut dec_buffer = [0u8; NOISE_MAX_SIZE];

                *len += try_ready!(io.poll_read(&mut enc_buffer[*len..]));
                if *len > 0 {
                    match noise.read_message(&enc_buffer[..*len], &mut dec_buffer) {
                        Ok(_) => self.next = Step::write(),
                        Err(e) => debug!("Error decoding message: {:?}", e),
                    };
                }
                if noise.is_handshake_finished() {
                    Ok(Async::Ready(self.wrap_up()))
                } else {
                    self.poll()
                }
```
`dec_buffer` is never used for anything.

So the question is does this Future allow for reading and understanding the application data?
In our case we're probably going to want to tell the other node during the handshake we're using `bam-json` or something so we know what codec to use afterwards.

So to properly calculate the size of the noise message you need to know:
1. Where you are in whatever handshake you're using e.g `s, se`, `e`
2. What curve you're using e.g 25519 (32 byte keys)
3. How much application data is meant to be sent at this point in the handshake.

If you get all these then you can figure out exactly how long the buffer should be. I think you could still keep your general structure -- probably just change the arrays in the enums to vecs.

We can leave the whole application data problem until we actually want to implement it. The question is, do you want to get rid of the 65535 and figure out how long each thing is meant to be (assuming 0 length application data) in the PR? We will have to do it eventually. The difficult thing about it won't be the actual calculation but adding the tracking the state of handshake you are in to your future. Can the snow session tell you this?

Reply:
I agree on point 1. Point 2 doesn't sound very useful.

I don't think it's over optimising to know how much data we need to read from a socket. That seems like something someone reading from a socket should know. In general I don't like throwing things into something and seeing if it's OK as a way of processing stuff.

For now I don't mind merging this and seeing how it works in practice. I'll review the rest of the code ASAP.

Reply:
Ok.. Should we open an issue for point 1 and another one to predict the size of the buffer?

Reply:
Please note that regarding prediction of the size of the buffer I think some of the work should go in the snow library itself.

Reply:
I think so too. I'd say, open an issue with the problem of "optimize buffer size during snow handshake". There, we can then reference that a contribution to snow is necessary and block the feature on that.

Reply:
#736 open